### PR TITLE
Fix invalid use of stack memory

### LINKED ISSUE #23689: utf8.cpp returns the address of a stack variable
https://github.com/dotnet/coreclr/blob/dffb3d30790131166c848283947125c59e428f38/src/pal/src/locale/utf8.cpp#L1242 returns the address of stack variable `bytes` via `byteUnknown`.

The variable is initialized here: https://github.com/dotnet/coreclr/blob/dffb3d30790131166c848283947125c59e428f38/src/pal/src/locale/utf8.cpp#L1174

 This looks like a bug porting the original code from C# to C.

cc @tmds

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Fixes #23689

### CODE REVIEW

File: src/pal/src/locale/utf8.cpp [hunk 1]
```diff
@@ -1168,24 +1168,23 @@ class UTF8Encoding
         return count;
     }
 
-    int GetBytesUnknown(BYTE* pSrc, int ch, BYTE **bytesUnknown)
+    int GetBytesUnknown(BYTE* pSrc, int ch, BYTE *bytesUnknown)
```
Comment (gfoidl) [inline src/pal/src/locale/utf8.cpp]:
```suggestion
    int GetBytesUnknown(BYTE* pSrc, int ch, BYTE* bytesUnknown)
```
to be consistent with other places.

**Comment by janvorli** [comment]
> we should consider updating the native implementation to have the same behavior.

I agree. When we have created a clone of the managed version for use in PAL, the idea was that the clone would be as close as possible to the managed version si that we could keep them in sync in the future.

**Comment by AaronRobinsonMSFT** [comment]
>When we have created a clone of the managed version for use in PAL

I personally think this should just be in native or in managed. I don't see the value in having two implementations. What am I missing?

**Comment by janvorli** [comment]
> I personally think this should just be in native or in managed. I don't see the value in having two implementations. What am I missing?

PAL cannot call managed code, but it still needs to perform conversions between UTF-16 and UTF-8, as the Windows APIs it emulates use UTF-16 but Unix APIs use UTF-8. And the PAL exists just on Unix, so we cannot use that code for managed code in general. But we would likely not want to pay the price of pinning and managed to native transition anyways.

**Pull request merged**